tests/phpunit/CRM/Event/Form/ParticipantTest.php

```diff
@@ -48,6 +48,23 @@ public function testSubmit(): void {
     $this->callAPISuccessGetSingle('Participant', ['id' => $form->getParticipantID()]);
   }
 
+  public function testSubmitWithExistingParticipant(): void {
+    $form = $this->getForm([], [
+      'register_date' => date('Ymd'),
+      'status_id' => 1,
+      'role_id' => [CRM_Core_PseudoConstant::getKey('CRM_Event_BAO_Participant', 'role_id', 'Attendee')],
+    ])->postProcess();
+    $this->assertEquals($this->getEventID(), $form->getEventID());
+    $this->callAPISuccessGetSingle('Participant', ['id' => $form->getParticipantID()]);
+    $participantID = $form->getParticipantID();
+
+    $form = new CRM_Event_Form_Participant();
+    $form->_action = CRM_Core_Action::ADD;
+    $form->_id = $participantID;
+    $form->_contactId = $this->getContactID();
+    $this->assertEquals(['event_id' => "This contact has already been assigned to this event."], $form->formRule(['event_id' => $this->getEventID()], [], $form));
+  }
```

Comment:
public function testSubmitWithExistingParticipant(): void {
    $submittedValues = [
      'register_date' => date('Ymd'),
      'status_id' => 1,
      'role_id' => [CRM_Core_PseudoConstant::getKey('CRM_Event_BAO_Participant', 'role_id', 'Attendee')],
    ];
    $form = $this->getForm([], $submittedValues)->postProcess();
    $this->assertEquals($this->getEventID(), $form->getEventID());
    $this->callAPISuccessGetSingle('Participant', ['id' => $form->getParticipantID()]);

    $this->getTestForm('CRM_Event_Form_Participant', $submittedValues + [
      'event_id' => $this->getEventID(),
      'contact_id' => $this->getContactID(),
    ], ['action' => 'add'])
      ->processForm(FormWrapper::VALIDATED);
    $this->assertValidationError(['event_id' => "This contact has already been assigned to this event."]);
  }